src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/FileBasedProgramsProjectSystem.cs

```diff
@@ -25,6 +27,9 @@ internal sealed class FileBasedProgramsProjectSystem : LanguageServerProjectLoad
     private readonly ILspServices _lspServices;
     private readonly ILogger<FileBasedProgramsProjectSystem> _logger;
     private readonly VirtualProjectXmlProvider _projectXmlProvider;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private CanonicalMiscFilesProject? _canonicalMiscFilesProject;
+    private readonly SemaphoreSlim _canonicalProjectGate = new(initialCount: 1);
```

Comment:
Just create the `CanonicalMiscFilesProject` instance when constructing this project system. Then have its internal operations use its internal gate. No need for a gate out here.